File: main/convert-bst-to-greater-tree/convert-bst-to-greater-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        def dfs(node):
            if node is not None:
                yield from dfs(node.right)
                yield node
                yield from dfs(node.left)
        
        acc = 0
        for node in dfs(root):
            acc += node.val
            node.val = acc
        
        dfs(root)
        return root

File: main/convert-bst-to-greater-tree/convert-bst-to-greater-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode convertBST(TreeNode root) {
        dfs(root, 0);
        return root;
    }
    
    private static int dfs(TreeNode root, int sum) {
        if (root == null) return sum;
        
        root.val += dfs(root.right, sum);
        return dfs(root.left, root.val);
    }
}
